### PR TITLE
FINERACT-1968: Advanced repayment allocation strategy - data model

### PR DESCRIPTION
**Pull request by b0c1**
FINERACT-1968: Advanced repayment allocation strategy - data model
- [x] m_loan_product_payment_allocation_rule table created
- [x] m_loan_payment_allocation_rule table created
- [x] LoanProductPaymentAllocationRule entity
- [x] LoanPaymentAllocationRule entity
- [x] List of enum converter to String converter to use `AllocationType` list

### CODE REVIEW

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanPaymentAllocationRule.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    @ManyToOne
+    @JoinColumn(name = "loan_id", nullable = false)
+    private Loan loan;
+
+    @Column(name = "transaction_type", nullable = false)
+    @Enumerated(EnumType.STRING)
+    private TransactionType transactionType;
+
+    @ElementCollection(fetch = FetchType.EAGER)
+    @Column(name = "allocation_type", nullable = false)
```
Comment (abraham-menyhart) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanPaymentAllocationRule.java]: In the 0122_add_payment_allocation_rule.xml you used `allocation_types` columnName, here is `allocation_type`, withouth `s`.

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProductPaymentAllocationRule.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    @ManyToOne
+    @JoinColumn(name = "loan_product_id", nullable = false)
+    private LoanProduct loanProduct;
+
+    @Column(name = "transaction_type", nullable = false)
+    @Enumerated(EnumType.STRING)
+    private TransactionType transactionType;
+
+    @ElementCollection(fetch = FetchType.EAGER)
+    @Column(name = "allocation_type", nullable = false)
```
Comment (abraham-menyhart) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProductPaymentAllocationRule.java]: In the 0122_add_payment_allocation_rule.xml you used `allocation_types` columnName, here is `allocation_type`, withouth `s.`

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanPaymentAllocationRule.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    private List<AllocationType> allocationTypes;
+
+    @Column(name = "future_installment_allocation_rule", nullable = false)
+    @Enumerated(EnumType.STRING)
+    private FutureInstallmentType futureInstallmentAllocationRule;
+
+    public Loan getLoan() {
```
Comment (abraham-menyhart) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanPaymentAllocationRule.java]: Why don't you use `@Getter`,`@Setter` annotation here?

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProductPaymentAllocationRule.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    private List<AllocationType> allocationTypes;
+
+    @Enumerated(EnumType.STRING)
+    @Column(name = "future_installment_allocation_rule", nullable = false)
+    private FutureInstallmentType futureInstallmentAllocationRule;
+
+    public LoanProduct getLoanProduct() {
```
Comment (abraham-menyhart) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProductPaymentAllocationRule.java]: Why don't you use `@Getter`,`@Setter` annotation here?

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/GenericEnumListConverter.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+ */
+package org.apache.fineract.portfolio.loanproduct.domain;
+
+import jakarta.persistence.AttributeConverter;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.commons.lang3.StringUtils;
+
+public abstract class GenericEnumListConverter<E extends Enum<E>> implements AttributeConverter<List<E>, String> {
```
Comment (taskain7) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/GenericEnumListConverter.java]: why do we have the same class in the core module and here too?

Comment (adamsaghy) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/GenericEnumListConverter.java]: Nice catch!

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanPaymentAllocationRule.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+import org.apache.fineract.portfolio.loanaccount.domain.Loan;
+import org.apache.fineract.portfolio.loanproduct.domain.FutureInstallmentAllocationRule;
+import org.apache.fineract.portfolio.loanproduct.domain.PaymentAllocationTransactionType;
+import org.apache.fineract.portfolio.loanproduct.domain.PaymentAllocationType;
+
+@Getter
+@Setter
+@Entity
+@Table(name = "m_loan_payment_allocation", uniqueConstraints = {
+        @UniqueConstraint(columnNames = { "loan_id", "transaction_type" }, name = "loan_payment_allocation_unique_transaction_type") })
```
Comment (abraham-menyhart) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanPaymentAllocationRule.java]: The unique constraint name defined here differs from the one in the xml.

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProductPaymentAllocationRule.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import org.apache.fineract.infrastructure.core.domain.AbstractAuditableWithUTCDateTimeCustom;
+
+@Getter
+@Setter
+@Entity
+@Table(name = "m_loan_product_payment_allocation", uniqueConstraints = { @UniqueConstraint(columnNames = { "loan_product_id",
+        "transaction_type" }, name = "loan_product_payment_allocation_unique_transaction_type") })
```
Comment (abraham-menyhart) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProductPaymentAllocationRule.java]: The unique constraint name defined here differs from the one in the xml.

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/PaymentAllocationType.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+    DUE_PRINCIPAL(DUE, PRINCIPAL),
+
+    DUE_INTEREST(DUE, INTEREST),
+
+    IN_ADVANCE_PENALTY(IN_ADVANCE, PENALTY),
+
+    IN_ADVANCE_FEE(IN_ADVANCE, FEE),
+
+    IN_ADVANCE_PRINCIPAL(IN_ADVANCE, PRINCIPAL);
```
Comment (abraham-menyhart) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/PaymentAllocationType.java]: IN_ADVANCE_INTEREST is missing.

**Review by abraham-menyhart — CHANGES_REQUESTED** [review]
Please see my comments

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanPaymentAllocationRule.java (hunk 1 above)
Comment (adamsaghy) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanPaymentAllocationRule.java]: Nice catch!

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProductPaymentAllocationRule.java (hunk 2 above)
Comment (adamsaghy) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProductPaymentAllocationRule.java]: Nice catch!

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/PaymentAllocationType.java (hunk 8 above)
Comment (adamsaghy) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/PaymentAllocationType.java]: Nice catch!

**Review by taskain7 — APPROVED** [review]
LGTM

**Review by abraham-menyhart — APPROVED** [review]
LGTM

**Pull request merged**